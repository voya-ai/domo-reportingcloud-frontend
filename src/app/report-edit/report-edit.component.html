<h2>Edit</h2>

<div *ngIf="isEditing else showEditButton">
  <div>
    <mat-form-field class="example-full-width">
      <input matInput placeholder="Report name" [(ngModel)]="report.name">
    </mat-form-field>
  </div>

  <div>
    <app-report-template-autocomplete
      (onSelectionChange)="templateSelectionChanged($event)"
      [templateName]="report.templateName">
    </app-report-template-autocomplete>
  </div>

  <div>
    <app-report-dataset-autocomplete
      (onSelectionChange)="datasetSelectionChanged($event)"
      [dataSet]="report.dataSet">
    </app-report-dataset-autocomplete>
  </div>

  <div *ngIf="columnsLoading">
    <mat-spinner></mat-spinner>
  </div>

  <div
    *ngIf="headerColumns.length > 0">
    <app-dataset-column-select
      [(columns)]="headerColumns"
      title="Header columns">
    </app-dataset-column-select>
  </div>

  <div
    *ngIf="contentColumns.length > 0">
    <app-dataset-column-select
      [(columns)]="contentColumns"
      title="Content columns">
    </app-dataset-column-select>
  </div>

  <button
    type="button"
    class="btn btn-primary"
    (click)="save()">Save</button>
</div>
<ng-template #showEditButton>
  <button
    type="button"
    class="btn btn-secondary"
    (click)="edit()">
    <i class="material-icons">
      edit
    </i>
  </button>
</ng-template>
