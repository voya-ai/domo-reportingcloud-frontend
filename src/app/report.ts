export interface Report {
  name?: string;
  slug: string;
  dataSet?: string;
  templateName?: string;
  headerColumns?: string[];
  contentColumns?: string[];
}
