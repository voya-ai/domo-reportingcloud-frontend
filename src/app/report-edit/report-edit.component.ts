import { Component, Input, Output, EventEmitter, OnInit } from '@angular/core';

import { DatasetService } from 'app/dataset.service';
import { ReportService } from 'app/report.service';

import { DatasetColumn } from 'app/dataset-column';
import { Report } from 'app/report';

import { Router } from '@angular/router';

@Component({
  selector: 'app-report-edit',
  templateUrl: './report-edit.component.html',
  styleUrls: ['./report-edit.component.css']
})
export class ReportEditComponent implements OnInit {
  @Input() report: Report;
  @Output() update : EventEmitter<Report> = new EventEmitter<Report>();

  headerColumns: DatasetColumn[] = [];
  contentColumns: DatasetColumn[] = [];

  isEditing: boolean = false;
  columnsLoading: boolean = false;

  constructor(private _datasetService: DatasetService,
    private _reportService: ReportService,
    private _router: Router)
  { }

  ngOnInit() {
  }

  datasetSelectionChanged(selection) {
    this.report.dataSet = selection.id;
    this.headerColumns = [];
    this.contentColumns = [];

    this.columnsLoading = true;

    var report = this.report;

    this._datasetService.getColumns(selection.id)
      .subscribe((columns : Array<DatasetColumn>) => {
        this.columnsLoading = false;

        // deep copy columns
        this.headerColumns = columns.map(function(column) {
          var result = { ...column };
          if(report.headerColumns.includes(result.name)) {
            result.checked = true;
          }
          return result;
        });
        this.contentColumns = columns.map(function(column) {
          var result = { ...column };
          if(report.contentColumns.includes(result.name)) {
            result.checked = true;
          }
          return result;
        });
      });
  }

  templateSelectionChanged(selection) {
    this.report.templateName = selection.templateName;
  }

  edit() {
    this.isEditing = true;
  }

  save(report) {
    this.report.headerColumns = this.headerColumns.filter(column => column.checked).map(column => column.name);
    this.report.contentColumns = this.contentColumns.filter(column => column.checked).map(column => column.name);
    this._reportService.update(this.report)
      .subscribe((report: Report) => {
        this.isEditing = false;
        this.update.emit(report);
      });
  }
}
