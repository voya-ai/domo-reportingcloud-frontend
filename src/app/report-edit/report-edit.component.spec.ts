import { async, ComponentFixture, TestBed } from '@angular/core/testing';

import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import {
  MatInputModule,
  MatProgressSpinnerModule
} from '@angular/material';

import { NoopAnimationsModule } from '@angular/platform-browser/animations';
import { RouterTestingModule } from '@angular/router/testing';

import { ReportService } from 'app/report.service';
import { DatasetService } from 'app/dataset.service';

import 'rxjs/add/observable/of';

import { MockComponent } from 'src/testing/component-stubs';

import { ReportEditComponent } from './report-edit.component';

class MockDatasetService {}
class MockReportService {}

describe('ReportEditComponent', () => {
  let component: ReportEditComponent;
  let fixture: ComponentFixture<ReportEditComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      providers: [
        { provide: DatasetService, useClass: MockDatasetService },
        { provide: ReportService, useClass: MockReportService }
      ],
      declarations: [
        ReportEditComponent,
        MockComponent({ selector: 'app-report-template-autocomplete', inputs: [ 'templateName'] }),
        MockComponent({ selector: 'app-report-dataset-autocomplete', inputs: [ 'dataSet'] }),
        MockComponent({ selector: 'app-dataset-column-select', inputs: [ 'columns', 'title' ] })
      ],
      imports: [
        NoopAnimationsModule,
        MatInputModule,
        MatProgressSpinnerModule,
        FormsModule,
        ReactiveFormsModule,
        RouterTestingModule.withRoutes([])
      ]
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(ReportEditComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should match the snapshot', () => {
    expect(fixture).toMatchSnapshot();
  });
});
