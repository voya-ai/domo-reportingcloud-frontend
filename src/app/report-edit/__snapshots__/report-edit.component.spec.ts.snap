// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReportEditComponent should match the snapshot 1`] = `
<app-report-edit
  _datasetService={[Function MockDatasetService]}
  _reportService={[Function MockReportService]}
  _router={[Function Router]}
  columnsLoading="false"
  contentColumns={[Function Array]}
  headerColumns={[Function Array]}
  isEditing="false"
  update={[Function EventEmitter]}
>
  <h2>
    Edit
  </h2>
  
  
    <button
    class="btn btn-secondary"
    type="button"
  >
    
      
    <i
      class="material-icons"
    >
      
        edit
      
    </i>
    
    
  </button>
  
  
  
</app-report-edit>
`;
